torch/_dynamo/backends/distributed.py

```diff
@@ -376,82 +380,34 @@ def forward(self, *args):
-            # to match what aot_autograd expects. See Note: [Fake Modules and AOTAutograd]
+            # We replace the submodules with lazy submodules which compile
+            # the corresponding submodules when they are run with real values
             def run_node(self, n: Node) -> Any:
-                args, kwargs = self.fetch_args_kwargs_from_env(n)
-                new_args = []
+                _, kwargs = self.fetch_args_kwargs_from_env(n)
                 assert fake_mode
-                for arg in args:
-                    if isinstance(arg, torch.Tensor) and not isinstance(
-                        arg, torch._subclasses.FakeTensor
-                    ):
-                        new_args.append(
-                            torch._dynamo.utils.to_fake_tensor(arg, fake_mode)
-                        )
-                    else:
-                        new_args.append(arg)
-
-                log.debug("run_node %s, %s got args %s", n.op, n.target, args_str(args))
-                assert isinstance(args, tuple)
-                assert isinstance(kwargs, dict)
 
                 if n.op == "call_module":
                     real_mod = self.fetch_attr(n.target)
-                    if fake_mode:
-                        curr_submod = deepcopy_to_fake_tensor(real_mod, fake_mode)
-                    else:
-                        curr_submod = real_mod
 
                     ddp_graph_log.debug(
-                        "\n---%s graph---\n%s", n.target, curr_submod.graph
+                        "\n---%s graph---\n%s", n.target, real_mod.graph
                     )
 
-                    # When calling the compiler on the submod, inputs (new_args) are expected to
-                    # be FakeTensors already since Dynamo would have made them FakeTensors in the
-                    # non-DDP flow.  However, the parameters are _not_ expected to be FakeTensors,
-                    # since this wrapping happens during compilation
-                    compiled_submod_real = self.compile_submod(
-                        real_mod, new_args, kwargs
-                    )
+                    assert len(kwargs) == 0, "We assume only args for these modules"
+                    compiled_submod_real = self.lazily_compile_submod(real_mod)
 
                     # We update the original (outer) graph with a call into the compiled module
                     # instead of the uncompiled one.
                     self.module.delete_submodule(n.target)
                     n.target = "compiled_" + n.target
                     self.module.add_submodule(n.target, compiled_submod_real)
 
-                    # Finally, we have to produce inputs for use compiling the next submodule,
-                    # and these need to be FakeTensors, so we execute the module under fake_mode
-                    with fake_mode:
-                        return curr_submod(*new_args, **kwargs)
-                else:
```

Comment:
how come this is removed?

Reply:
We don't need the example outputs when running `SubmodCompiler` anymore.

We just need to replace the submodules with lazily-compiling submodules.

Reply:
oh i see. we no longer care about propagating 'args' during this traversal, so i guess we are propagating 'None' value implicitly instead of propagating real or fake values like we used to?

If so, please add a comment. (but i think that's fine).

Reply:
Yes let me double check if the behaviour when returning None in the `run_node` is to fallback on that node's hard-coded args. Maybe the kwargs checking needs to be deferred to the lazy submodule if not.

Reply:
Actually, kwargs checking does not need to be deferred. If there are any kwargs, we will still error, since even if the kwargs value are `None`, the node would still have been passed a kwarg, which we do not currently support.